src/submodule.rs

```diff
@@ -33,6 +33,23 @@ impl<'repo> Submodule<'repo> {
         unsafe { crate::opt_bytes(self, raw::git_submodule_branch(self.raw)) }
     }
 
+    ///
+    pub fn clone(
+        &mut self,
+        repo: Option<&mut &mut Repository>,
+        opts: &mut SubmoduleUpdateOptions<'_>,
```

Comment:
There is probably a nicer way to create the public API but I have zero experience with raw bindings :(

Reply:
Instead of passing the `repo` parameter, it should return a `Repository` as a return value.  It should look roughly the same as [`Submodule::open`](https://github.com/rust-lang/git2-rs/blob/c2a26133731d152f67a4fb11b35c4e139441e60c/src/submodule.rs#L113-L123), plus the `opts` parameter.

You can grab the docs from the [header file](https://github.com/libgit2/libgit2/blob/c71321a099373753c22055921c8f538afed5ebb6/include/git2/submodule.h#L289-L299), and just adjust them similar to how the other methods do.

It would also be nice to have a test for it. There are a few at the bottom of the file, and you can probably just copy one of those.

Reply:
That's what I wasn't sure off. I haven't really touched any C code in my life so it's quite hard to get to know the calling conventions. I thought that it was an optional parameter that was needed to function.

https://libgit2.org/libgit2/#HEAD/group/submodule/git_submodule_clone

But after re-reading it seems that it is actually the output value.